<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title></title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:conda@f47048a8fffe.(none)" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#CONTACT">CONTACT</a></li>
  <li><a href="#USAGE">USAGE</a></li>
  <li><a href="#OPTIONS">OPTIONS</a>
    <ul>
      <li><a href="#Fuctions">*********************************************************** Fuctions ***********************************************************</a></li>
      <li><a href="#Global-Options">******************************************************** Global Options ********************************************************</a></li>
      <li><a href="#Local-Options">******************************************************** Local Options *********************************************************</a>
        <ul>
          <li><a href="#Options-of---Assemble-for-reads-assembly-gene-prediction-and-annotation">======================= Options of &quot;--Assemble&quot; for reads assembly, gene prediction and annotation ============================</a></li>
          <li><a href="#Options-for---CoreTree-constructing">======================================== Options for &quot;--CoreTree&quot; constructing ================================================</a></li>
          <li><a href="#Options-for---Pan-analysis">===================================== Options for &quot;--Pan&quot; analysis ============================================================</a></li>
          <li><a href="#Options-for---orthoF-analysis">===================================== Options for &quot;--orthoF&quot; analysis =========================================================</a></li>
          <li><a href="#Options-for---ANI-analysis">===================================== Options for &quot;--ANI&quot; analysis ============================================================</a></li>
          <li><a href="#Options-for---VAR-analysis">===================================== Options for &quot;--VAR&quot; analysis ============================================================</a></li>
        </ul>
      </li>
      <li><a href="#Paths-of-external-programs">************************************* Paths of external programs ***************************************************************</a></li>
    </ul>
  </li>
  <li><a href="#CODES">CODES</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>PGCGAP</p>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>The prokaryotic genomics and comparative genomics analysis pipeline</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Hualin Liu</p>

<h1 id="CONTACT">CONTACT</h1>

<p>liaochenlanruo@webmail.hzau.edu.cn</p>

<h1 id="USAGE">USAGE</h1>

<pre><code>  $ pgcgap [Fuctions] [Options]

  example 1: Perform all functions, take the bacillus thuringiensis as an example, total 4 strains for analysis

             For the sake of flexibility, The &quot;VAR&quot; function needs to be added additionally

             pgcgap --All --ReadsPath &lt;PATH&gt; --reads1 .R1.clean.fastq.gz --reads2 .R2.clean.fastq.gz --suffix_len &lt;INT&gt; --kmmer 81 --genus bacillus --species thuringiensis --codon 11 --strain_num 4 --threads 4 --VAR --refgbk &lt;FILE&gt; --qualtype &lt;STRING&gt;

  example 2: Conduct reads assembly, gene prediction and annotation

             pgcgap --Assemble --ReadsPath &lt;PATH&gt; --reads1 .R1.clean.fastq.gz --reads2 .R2.clean.fastq.gz --suffix_len &lt;INT&gt; --kmmer 81 --genus bacillus --species thuringiensis --codon 11 --threads 4

  example 3: Constructing the phylogenetic trees of single-core proteins and core SNPs

             pgcgap --CoreTree --CDsPath &lt;PATH&gt; --AAsPath &lt;PATH&gt; --codon 11 --strain_num 4 --threads 4

  example 4: Conduct pan-genome analysis

             pgcgap --Pan --codon 11 --threads 4 --GffPath &lt;PATH&gt;

  example 5: Inference of orthologous gene groups

             pgcgap --orthoF --threads 4 --AAsPath &lt;PATH&gt;

  example 6: Compute whole-genome Average Nucleotide Identity (ANI)

             pgcgap --ANI --threads 4 --queryL &lt;FILE&gt; --refL &lt;FILE&gt; --ANIO &lt;FILE&gt; --Scaf_suffix &lt;STRING&gt;

  example 7: Run COG annotation for each strain

              pgcgap --COG --threads 4 --AAsPath &lt;PATH&gt;

  example 8: Varients calling and phylogenetic tree construction based on reference genome

             pgcgap --VAR --threads 4 --refgbk &lt;FILE&gt; --ReadsPath &lt;PATH&gt; --reads1 .R1.clean.fastq.gz --reads2 .R2.clean.fastq.gz --suffix_len &lt;INT&gt; --strain_num &lt;INT&gt; --qualtype &lt;STRING&gt; </code></pre>

<h1 id="OPTIONS">OPTIONS</h1>

<dl>

<dt id="help"><b>[--help]</b></dt>
<dd>

<p>Print the help message and exit</p>

</dd>
</dl>

<dl>

<dt id="version"><b>[--version]</b></dt>
<dd>

<p>Show version number of PGCGAP and exit</p>

</dd>
</dl>

<dl>

<dt id="check-external-programs"><b>[--check-external-programs]</b></dt>
<dd>

<p>Check if all of the required external programs can be found and are executable, then exit</p>

</dd>
</dl>

<dl>

<dt id="setup-COGdb"><b>[--setup-COGdb]</b></dt>
<dd>

<p>Setup COG database. Users should execute &quot;pgcgap --setup-COGdb&quot; after first installation of pgcgap</p>

</dd>
</dl>

<h2 id="Fuctions">*********************************************************** Fuctions ***********************************************************</h2>

<dl>

<dt id="All"><b>[--All]</b></dt>
<dd>

<p>Perform Assemble, CoreTree, Pan, OrthoF, ANI and COG functions with one command</p>

</dd>
</dl>

<dl>

<dt id="Assemble"><b>[--Assemble]</b></dt>
<dd>

<p>Assemble reads ( paired-end only ) into contigs, predict genes and annotate them</p>

</dd>
</dl>

<dl>

<dt id="CoreTree"><b>[--CoreTree]</b></dt>
<dd>

<p>Construct single-core proteins tree and SNPs tree of single core genes</p>

</dd>
</dl>

<dl>

<dt id="Pan"><b>[--Pan]</b></dt>
<dd>

<p>Run &quot;roary&quot; pan genome pipeline with gff3 files</p>

</dd>
</dl>

<dl>

<dt id="OrthoF"><b>[--OrthoF]</b></dt>
<dd>

<p>Identify orthologous protein sequence families with &quot;OrthoFinder&quot;</p>

</dd>
</dl>

<dl>

<dt id="ANI"><b>[--ANI]</b></dt>
<dd>

<p>Compute whole-genome Average Nucleotide Identity ( ANI )</p>

</dd>
</dl>

<dl>

<dt id="COG"><b>[--COG]</b></dt>
<dd>

<p>Run COG annotation for each strain (*.faa), and generate a table containing the relative abundance of each flag for all strains</p>

</dd>
</dl>

<dl>

<dt id="VAR"><b>[--VAR]</b></dt>
<dd>

<p>Rapid haploid variant calling and core genome alignment with &quot;Snippy&quot;</p>

</dd>
</dl>

<h2 id="Global-Options">******************************************************** Global Options ********************************************************</h2>

<dl>

<dt id="strain_num-INT"><b>[--strain_num (INT)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--CoreTree&quot; and &quot;--VAR&quot;]</i> The total number of strains used for analysis, including reference genomes</p>

</dd>
</dl>

<dl>

<dt id="ReadsPath-PATH"><b>[--ReadsPath (PATH)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--Assemble&quot; and &quot;--VAR&quot;]</i> Reads of all strains as file paths ( Default ./Reads )</p>

</dd>
</dl>

<dl>

<dt id="AAsPath-PATH"><b>[--AAsPath (PATH)]</b></dt>
<dd>

<p><i>[Required by &quot;--CoreTree&quot; and &quot;--COG&quot;]</i> Amino acids of all strains as fasta file paths, ( Default &quot;./Results/Annotations/AAs&quot; )</p>

</dd>
</dl>

<dl>

<dt id="reads1-STRING"><b>[--reads1 (STRING)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--Assemble&quot; and &quot;--VAR&quot;]</i> The suffix name of reads 1 ( for example: if the name of reads 1 is &quot;YBT-1520_L1_I050.R1.clean.fastq.gz&quot;, &quot;YBT-1520&quot; is the strain same, so the suffix name should be &quot;.R1.clean.fastq.gz&quot; )</p>

</dd>
</dl>

<dl>

<dt id="reads2-STRING"><b>[--reads2 (STRING)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--Assemble&quot; and &quot;--VAR&quot;]</i> The suffix name of reads 2( for example: if the name of reads 2 is &quot;YBT-1520_2.fq&quot;, the suffix name should be _2.fq&quot; )</p>

</dd>
</dl>

<dl>

<dt id="codon-INT"><b>[--codon (INT)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--Assemble&quot;, &quot;--CoreTree&quot; and &quot;--Pan&quot;]</i> Translation table ( Default 11 )</p>

</dd>
</dl>

<ul>

<pre><code> 1   Universal code
 2   Vertebrate mitochondrial code
 3   Yeast mitochondrial code
 4   Mold, Protozoan, and Coelenterate Mitochondrial code and Mycoplasma/Spiroplasma code
 5   Invertebrate mitochondrial
 6   Ciliate, Dasycladacean and Hexamita nuclear code
 9   Echinoderm and Flatworm mitochondrial code
 10  Euplotid nuclear code
 11  Bacterial, archaeal and plant plastid code ( Default )
 12  Alternative yeast nuclear code
 13  Ascidian mitochondrial code
 14  Alternative flatworm mitochondrial code
 15  Blepharisma nuclear code
 16  Chlorophycean mitochondrial code
 21  Trematode mitochondrial code
 22  Scenedesmus obliquus mitochondrial code
 23  Thraustochytrium mitochondrial code</code></pre>

</ul>

<dl>

<dt id="suffix_len-INT"><b>[--suffix_len (INT)]</b></dt>
<dd>

<p><i>[Required by &quot;--All&quot;, &quot;--Assemble&quot; and &quot;--VAR&quot;]</i> <b>(Strongly recommended)</b> The suffix length of the reads, that is the length of your reads name minus the length of your strain name. For example the --suffix_len of &quot;YBT-1520_L1_I050.R1.clean.fastq.gz&quot; is 26 ( &quot;YBT-1520&quot; is the strain name ) ( Default 0 )</p>

</dd>
</dl>

<dl>

<dt id="logs-STRING"><b>[--logs (STRING)]</b></dt>
<dd>

<p>Name of the log file ( Default Logs.txt )</p>

</dd>
</dl>

<dl>

<dt id="threads-INT"><b>[--threads (INT)]</b></dt>
<dd>

<p>Number of threads to be used ( Default 4 )</p>

</dd>
</dl>

<h2 id="Local-Options">******************************************************** Local Options *********************************************************</h2>

<h3 id="Options-of---Assemble-for-reads-assembly-gene-prediction-and-annotation">======================= Options of &quot;--Assemble&quot; for reads assembly, gene prediction and annotation ============================</h3>

If you use the results of "--Assemble" function in your work, please also cite:

</br>

</br>Shaun D Jackman, Benjamin P Vandervalk, Hamid Mohamadi, Justin Chu, Sarah Yeo, S Austin Hammond, Golnaz Jahesh, Hamza Khan, Lauren Coombe, Ren&eacute; L Warren, and Inanc Birol (2017). ABySS 2.0: Resource-efficient assembly of large genomes using a Bloom filter. Genome research, 27(5), 768-777. doi:10.1101/gr.214346.116

</br>

</br>Seemann T. Prokka: rapid prokaryotic genome annotation. Bioinformatics 2014 Jul 15;30(14):2068-9. PMID:24642063

<dl>

<dt id="kmmer-INT"><b>[--kmmer (INT)]</b></dt>
<dd>

<p><i>[Required]</i> k-mer size for genome assembly ( Default 81 )</p>

</dd>
</dl>

<dl>

<dt id="genus-STRING"><b>[--genus (STRING)]</b></dt>
<dd>

<p>Genus name of your strain ( Default &quot;NA&quot; )</p>

</dd>
</dl>

<dl>

<dt id="species-STRING"><b>[--species (STRING)]</b></dt>
<dd>

<p>Species name of your strain ( Default &quot;NA&quot; )</p>

</dd>
</dl>

<h3 id="Options-for---CoreTree-constructing">======================================== Options for &quot;--CoreTree&quot; constructing ================================================</h3>

If you use the results of "--CoreTree" function in your work, please also cite:

</br>

</br>CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659

</br>

</br>CD-HIT: accelerated for clustering the next generation sequencing data", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152

</br>

</br>Katoh K, Misawa K, Kuma K, Miyata T. MAFFT: a novel method for rapid multiple sequence alignment based on fast Fourier transform. Nucleic Acids Res. 2002;30(14):3059-3066

</br>

</br>Price MN, Dehal PS, Arkin AP. FastTree 2--approximately maximum-likelihood trees for large alignments. PLoS One. 2010;5(3):e9490. Published 2010 Mar 10. doi:10.1371/journal.pone.0009490

</br>

</br>Mikita Suyama, David Torrents, and Peer Bork (2006) PAL2NAL: robust conversion of protein sequence alignments into the corresponding codon alignments. Nucleic Acids Res. 34, W609-W612

</br>

</br>"SNP-sites: rapid efficient extraction of SNPs from multi-FASTA alignments", Andrew J. Page, Ben Taylor, Aidan J. Delaney, Jorge Soares, Torsten Seemann, Jacqueline A. Keane, Simon R. Harris, Microbial Genomics 2(4), (2016)

</br>

</br>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R. "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". Nucleic Acids Res. 2015 Feb 18;43(3):e15. doi: 10.1093/nar/gku1196

<dl>

<dt id="CDsPath-PATH"><b>[--CDsPath (PATH)]</b></dt>
<dd>

<p><i>[Required]</i> CDs of all strains as fasta file paths, ( Default &quot;./Results/Annotations/CDs&quot; )</p>

</dd>
</dl>

<dl>

<dt id="c-FLOAT"><b>[-c (FLOAT)]</b></dt>
<dd>

<p>Sequence identity threshold, ( Default 0.5)</p>

</dd>
</dl>

<dl>

<dt id="n-INT"><b>[-n (INT)]</b></dt>
<dd>

<p>Word_length, see user&#39;s guide of CD-HIT for choosing it ( Default 2 )</p>

</dd>
</dl>

<dl>

<dt id="G-INT"><b>[-G (INT)]</b></dt>
<dd>

<p>Use global (set to 1) or local (set to 0) sequence identity, ( Default 0 )</p>

</dd>
</dl>

<dl>

<dt id="t-INT"><b>[-t (INT)]</b></dt>
<dd>

<p>Tolerance for redundance ( Default 0 )</p>

</dd>
</dl>

<dl>

<dt id="aL-FLOAT"><b>[-aL (FLOAT)]</b></dt>
<dd>

<p>Alignment coverage for the longer sequence. If set to 0.9, the alignment must covers 90% of the sequence ( Default 0.9 )</p>

</dd>
</dl>

<dl>

<dt id="aS-FLOAT"><b>[-aS (FLOAT)]</b></dt>
<dd>

<p>Alignment coverage for the shorter sequence. If set to 0.9, the alignment must covers 90% of the sequence ( Default 0.9 )</p>

</dd>
</dl>

<dl>

<dt id="g-INT"><b>[-g (INT)]</b></dt>
<dd>

<p>If set to 0, a sequence is clustered to the first cluster that meet the threshold (fast cluster). If set to 1, the program will cluster it into the most similar cluster that meet the threshold (accurate but slow mode, Default 1)</p>

</dd>
</dl>

<dl>

<dt id="d-INT"><b>[-d (INT)]</b></dt>
<dd>

<p>length of description in .clstr file. if set to 0, it takes the fasta defline and stops at first space ( Default 0 )</p>

</dd>
</dl>

<h3 id="Options-for---Pan-analysis">===================================== Options for &quot;--Pan&quot; analysis ============================================================</h3>

If you use the results of "--Pan" function in your work, please also cite:

</br>

</br>"Roary: Rapid large-scale prokaryote pan genome analysis", Andrew J. Page, Carla A. Cummins, Martin Hunt, Vanessa K. Wong, Sandra Reuter, Matthew T. G. Holden, Maria Fookes, Daniel Falush, Jacqueline A. Keane, Julian Parkhill, Bioinformatics, (2015). doi: http://dx.doi.org/10.1093/bioinformatics/btv421

<dl>

<dt id="GffPath-PATH"><b>[--GffPath (PATH)]</b></dt>
<dd>

<p><i>[Required]</i> Gff files of all strains as paths ( Default &quot;./Results/Annotations/GFF&quot; )</p>

</dd>
</dl>

<h3 id="Options-for---orthoF-analysis">===================================== Options for &quot;--orthoF&quot; analysis =========================================================</h3>

If you use the results of "--orthoF" function in your work, please also cite:

</br>

</br>Emms, D.M. and Kelly, S. (2018) OrthoFinder2: fast and accurate phylogenomic orthology analysis from gene sequences. bioRxiv

<dl>

<dt id="Sprogram-STRING"><b>[--Sprogram (STRING)]</b></dt>
<dd>

<p>Sequence search program, Options: blast, mmseqs, blast_gz, diamond ( Default blast )</p>

</dd>
</dl>

<h3 id="Options-for---ANI-analysis">===================================== Options for &quot;--ANI&quot; analysis ============================================================</h3>

If you use the results of "--ANI" function in your work, please also cite:

</br>

</br>Jain C, Rodriguez-R LM, Phillippy AM, Konstantinidis KT, Aluru S. High throughput ANI analysis of 90K prokaryotic genomes reveals clear species boundaries. Nat Commun. 2018;9(1):5114. Published 2018 Nov 30. doi:10.1038/s41467-018-07641-9

<dl>

<dt id="queryL-FILE"><b>[--queryL (FILE)]</b></dt>
<dd>

<p><i>[Required]</i> The file containing paths to query genomes, one per line ( Default scaf.list )</p>

</dd>
</dl>

<dl>

<dt id="refL-FILE"><b>[--refL (FILE)]</b></dt>
<dd>

<p><i>[Required]</i> The file containing paths to reference genomes, one per line. ( Default scaf.list )</p>

</dd>
</dl>

<dl>

<dt id="ANIO-FILE"><b>[--ANIO (FILE)]</b></dt>
<dd>

<p>The name of output file ( Default &quot;Results/ANI/ANIs&quot; )</p>

</dd>
</dl>

<dl>

<dt id="Scaf_suffix-STRING"><b>[--Scaf_suffix (STRING)]</b></dt>
<dd>

<p>The suffix of scaffolds or genomes ( Default &quot;-8.fa&quot; )</p>

</dd>
</dl>

<h3 id="Options-for---VAR-analysis">===================================== Options for &quot;--VAR&quot; analysis ============================================================</h3>

If you use the results of "--VAR" function in your work, please also cite:

</br>

</br>Joshi NA, Fass JN. (2011). Sickle: A sliding-window, adaptive, quality-based trimming tool for FastQ files (Version 1.33) [Software].  Available at https://github.com/najoshi/sickle.

</br>

</br>Seemann T (2015) snippy: fast bacterial variant calling from NGS reads https://github.com/tseemann/snippy.

</br>

</br>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R. "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". Nucleic Acids Res. 2015 Feb 18;43(3):e15. doi: 10.1093/nar/gku1196

</br>

<dl>

<dt id="refgbk-FILE"><b>[--refgbk (FILE)]</b></dt>
<dd>

<p><i>[Required]</i> The <b>full path and name</b> of reference genome in GENBANK format ( <b>recommended</b> ), fasta format is also OK. For example: &quot;/mnt/g/test/ref.gbk&quot;</p>

</dd>
</dl>

<dl>

<dt id="qualtype-STRING"><b>[--qualtype (STRING)]</b></dt>
<dd>

<p><i>[Required]</i> Type of quality values (solexa (CASAVA &lt; 1.3), illumina (CASAVA 1.3 to 1.7), sanger (which is CASAVA &gt;= 1.8)). ( Default sanger )</p>

</dd>
</dl>

<dl>

<dt id="qual-INT"><b>[--qual (INT)]</b></dt>
<dd>

<p>Threshold for trimming based on average quality in a window. ( Default 20 )</p>

</dd>
</dl>

<dl>

<dt id="length-INT"><b>[--length (INT)]</b></dt>
<dd>

<p>Threshold to keep a read based on length after trimming. ( Default 20 )</p>

</dd>
</dl>

<dl>

<dt id="mincov-INT"><b>[--mincov (INT)]</b></dt>
<dd>

<p>The minimum number of reads covering a site to be considered ( Default 10 )</p>

</dd>
</dl>

<dl>

<dt id="minfrac-FLOAT"><b>[--minfrac (FLOAT)]</b></dt>
<dd>

<p>The minimum proportion of those reads which must differ from the reference ( Default 0.9 )</p>

</dd>
</dl>

<dl>

<dt id="minqual-INT"><b>[--minqual (INT)]</b></dt>
<dd>

<p>The minimum VCF variant call &quot;quality&quot; ( Default 100 )</p>

</dd>
</dl>

<dl>

<dt id="ram-INT"><b>[--ram (INT)]</b></dt>
<dd>

<p>Try and keep RAM under this many GB ( Default 8 )</p>

</dd>
</dl>

<dl>

<dt id="tree_builder-STRING"><b>[--tree_builder (STRING)]</b></dt>
<dd>

<p>Application to use for tree building [raxml|fasttree|hybrid] ( Default fasttree )</p>

</dd>
</dl>

<dl>

<dt id="iterations-INT"><b>[--iterations (INT)]</b></dt>
<dd>

<p>Maximum No. of iterations for gubbins ( Default 5 )</p>

</dd>
</dl>

<h2 id="Paths-of-external-programs">************************************* Paths of external programs ***************************************************************</h2>

<ul>

<p>Not needed if they were in the environment variables path. Users can check with the &quot;--check-external-programs&quot; option for the essential programs</p>

</ul>

<dl>

<dt id="abyss-bin-PATH"><b>[--abyss-bin (PATH)]</b></dt>
<dd>

<p>Path to abyss binary file. Default tries if abyss is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="prodigal-bin-PATH"><b>[--prodigal-bin (PATH)]</b></dt>
<dd>

<p>Path to prodigal binary file. Default tries if prodigal is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="prokka-bin-PATH"><b>[--prokka-bin (PATH)]</b></dt>
<dd>

<p>Path to prokka binary file. Default tries if prokka is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="cd-hit-bin-PATH"><b>[--cd-hit-bin (PATH)]</b></dt>
<dd>

<p>Path to cd-hit binary file. Default tries if cd-hit is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="mafft-bin-PATH"><b>[--mafft-bin (PATH)]</b></dt>
<dd>

<p>Path to mafft binary file. Default tries if mafft is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="fasttree-bin-PATH"><b>[--fasttree-bin (PATH)]</b></dt>
<dd>

<p>Path to the fasttree binary file. Default tries if fasttree is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="pal2nal-bin-PATH"><b>[--pal2nal-bin (PATH)]</b></dt>
<dd>

<p>Path to the pal2nal.pl binary file. Default tries if pal2nal.pl is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="snp-sites-bin-PATH"><b>[--snp-sites-bin (PATH)]</b></dt>
<dd>

<p>Path to the snp-sites binary file. Default tries if snp-sites is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="roary-bin-PATH"><b>[--roary-bin (PATH)]</b></dt>
<dd>

<p>Path to the roary binary file. Default tries if roary is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="orthofinder-bin-PATH"><b>[--orthofinder-bin (PATH)]</b></dt>
<dd>

<p>Path to the orthofinder binary file. Default tries if orthofinder is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="fastANI-bin-PATH"><b>[--fastANI-bin (PATH)]</b></dt>
<dd>

<p>Path to the fastANI binary file. Default tries if fastANI is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="gubbins-bin-PATH"><b>[--gubbins-bin (PATH)]</b></dt>
<dd>

<p>Path to the run_gubbins.py binary file. Default tries if run_gubbins.py is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="snippy-bin-PATH"><b>[--snippy-bin (PATH)]</b></dt>
<dd>

<p>Path to the snippy binary file. Default tries if snippy is in PATH;</p>

</dd>
</dl>

<dl>

<dt id="sickle-bin-PATH"><b>[--sickle-bin (PATH)]</b></dt>
<dd>

<p>Path to the sickle-trim binary file. Default tries if sickle is in PATH;</p>

</dd>
</dl>

<h1 id="CODES">CODES</h1>


</body>

</html>


